Build failure: qt6.qtgrpc

### Steps To Reproduce

Steps to reproduce the behavior:
1. build qt6.full on nixos-unstable channel

### Build log

```
@nix { "action": "setPhase", "phase": "qtPreHook" }
qtPreHook
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/qf242hz9q7z07kvl92l6vcrvs4y0lmy4-qtgrpc-everywhere-src-6.5.2.tar.xz
source root is qtgrpc-everywhere-src-6.5.2
setting SOURCE_DATE_EPOCH to timestamp 1688780772 of file qtgrpc-everywhere-src-6.5.2/.QT-ENTERPRISE-LICENSE-AGREEMENT
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
fixing cmake files...
cmake flags: -GNinja -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALE>
-- The CXX compiler identification is GNU 12.3.0
-- The C compiler identification is GNU 12.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/75slks1wr3b3sxr5advswjzg9lvbv9jc-gcc-wrapper-12.3.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/75slks1wr3b3sxr5advswjzg9lvbv9jc-gcc-wrapper-12.3.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Found OpenGL: /nix/store/x4jnkjc5229rawl7kpdhnmyy25y67ssb-libGL-1.6.0/lib/libOpenGL.so
-- Found WrapOpenGL: TRUE
-- Found XKB: /nix/store/gj8rdi3zrgvxxdrpvpffjql8kb3cm9lj-libxkbcommon-1.5.0/lib/libxkbcommon.so (Required is at least version "0.5.0")
-- Found WrapVulkanHeaders: /nix/store/7gg3vx5ghqk5r49j6zlmm4w581vakd06-vulkan-headers-1.3.261/include

-- Configuration summary shown below. It has also been written to /build/qtgrpc-everywhere-src-6.5.2/build/config.summary
-- Configure with --log-level=STATUS or higher to increase CMake's message verbosity. The log level does not persist across reconfigurations.

-- Configure summary:

Qt Protobuf tools:
  Qt Protobuf generator .................. yes
Qt GRPC:
  gRPC support ........................... yes
  Native gRPC support .................... yes
Qt GRPC tools:
  Qt GRPC generator ...................... yes


-- Configuring done (2.5s)
-- Generating done (0.0s)
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_POLICY_DEFAULT_CMP0025


-- Build files have been written to: /build/qtgrpc-everywhere-src-6.5.2/build
cmake: enabled parallel building
cmake: enabled parallel installing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j16
[1/63] Generating ../../mkspecs/modules/qt_lib_protobuf_private.pri
[2/63] Generating ../../mkspecs/modules/qt_lib_grpc_private.pri
[3/63] Running syncqt.cpp for module: QtProtobuf
[4/63] Running syncqt.cpp for module: QtGrpc
[5/63] Generating version linker script for target Grpc
[6/63] Generating version linker script for target Protobuf
[7/63] Building CXX object src/tools/qtgrpcgen/CMakeFiles/qtgrpcgen.dir/grpctemplates.cpp.o
[8/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/generatorbase.cpp.o
FAILED: src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/generatorbase.cpp.o
/nix/store/75slks1wr3b3sxr5advswjzg9lvbv9jc-gcc-wrapper-12.3.0/bin/g++ -DPROTOBUF_USE_DLLS -I/build/qtgrpc-everywhere-src-6.5.2/src/tools/qtprotoccommon -DNDEBUG -O2 -std=gnu++17 -fPIC -Wall -Wextra -MD -MT src/>
/build/qtgrpc-everywhere-src-6.5.2/src/tools/qtprotoccommon/generatorbase.cpp:8:10: fatal error: google/protobuf/stubs/logging.h: No such file or directory
    8 | #include <google/protobuf/stubs/logging.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[9/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/qprotobufgenerator.cpp.o
FAILED: src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/qprotobufgenerator.cpp.o
/nix/store/75slks1wr3b3sxr5advswjzg9lvbv9jc-gcc-wrapper-12.3.0/bin/g++ -DPROTOBUF_USE_DLLS -DQT_DISABLE_DEPRECATED_UP_TO=0x050000 -DQT_EXPLICIT_QFILE_CONSTRUCTION_FROM_PATH -DQT_LEAN_HEADERS=1 -DQT_NO_AS_CONST=1>
/build/qtgrpc-everywhere-src-6.5.2/src/tools/qtprotobufgen/qprotobufgenerator.cpp:20:10: fatal error: google/protobuf/stubs/logging.h: No such file or directory
   20 | #include <google/protobuf/stubs/logging.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[10/63] Building CXX object src/tools/qtgrpcgen/CMakeFiles/qtgrpcgen.dir/qgrpcgenerator.cpp.o
FAILED: src/tools/qtgrpcgen/CMakeFiles/qtgrpcgen.dir/qgrpcgenerator.cpp.o
/nix/store/75slks1wr3b3sxr5advswjzg9lvbv9jc-gcc-wrapper-12.3.0/bin/g++ -DPROTOBUF_USE_DLLS -DQT_DISABLE_DEPRECATED_UP_TO=0x050000 -DQT_EXPLICIT_QFILE_CONSTRUCTION_FROM_PATH -DQT_LEAN_HEADERS=1 -DQT_NO_AS_CONST=1>
/build/qtgrpc-everywhere-src-6.5.2/src/tools/qtgrpcgen/qgrpcgenerator.cpp:15:10: fatal error: google/protobuf/stubs/logging.h: No such file or directory
   15 | #include <google/protobuf/stubs/logging.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[11/63] Automatic MOC for target Protobuf
[12/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/options.cpp.o
[13/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/utils.cpp.o
[14/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/commontemplates.cpp.o
[15/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/main.cpp.o
[16/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/enumdefinitionprinter.cpp.o
[17/63] Building CXX object src/tools/qtgrpcgen/CMakeFiles/qtgrpcgen.dir/main.cpp.o
[18/63] Building CXX object src/tools/qtgrpcgen/CMakeFiles/qtgrpcgen.dir/serverdeclarationprinter.cpp.o
[19/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/baseprinter.cpp.o
[20/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/enumdeclarationprinter.cpp.o
[21/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/messagedefinitionprinter.cpp.o
[22/63] Building CXX object src/tools/qtprotoccommon/CMakeFiles/QtProtocCommon.dir/generatorcommon.cpp.o
[23/63] Building CXX object src/tools/qtprotobufgen/CMakeFiles/qtprotobufgen.dir/messagedeclarationprinter.cpp.o
ninja: build stopped: subcommand failed.
```

### Additional context

Add any other context about the problem here.

### Notify maintainers
@NickCao 
@milahu 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.42, NixOS, 23.05 (Stoat), 23.05.20230801.bd836ac`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/libraries/qt-6/default.nix
+++ b/pkgs/development/libraries/qt-6/default.nix
@@ -3,6 +3,7 @@
 , stdenv
 , fetchurl
 , fetchpatch
+, fetchpatch2
 , makeSetupHook
 , makeWrapper
 , gst_all_1
@@ -104,7 +105,15 @@ let
       qtdatavis3d = callPackage ./modules/qtdatavis3d.nix { };
       qtdeclarative = callPackage ./modules/qtdeclarative.nix { };
       qtdoc = callPackage ./modules/qtdoc.nix { };
-      qtgrpc = callPackage ./modules/qtgrpc.nix { };
+      qtgrpc = callPackage ./modules/qtgrpc.nix {
+        patches = [
+          (fetchpatch2 {
+            # fix compatibility with protobuf 23
+            url = "https://gitlab.archlinux.org/archlinux/packaging/packages/qt6-grpc/-/raw/d6b33bd915dc6e63b30db2cd29150d55b289d7ed/protobuf-23.patch";
+            hash = "sha256-KQAcrjQ3rK9pDQUOUK6AS4ej8YvtRv9WZOxby31Y5r4=";
+          })
+        ];
+      };
       qthttpserver = callPackage ./modules/qthttpserver.nix { };
       qtimageformats = callPackage ./modules/qtimageformats.nix { };
       qtlanguageserver = callPackage ./modules/qtlanguageserver.nix { };

--- a/pkgs/development/libraries/qt-6/modules/qtgrpc.nix
+++ b/pkgs/development/libraries/qt-6/modules/qtgrpc.nix
@@ -3,10 +3,12 @@
 , qtdeclarative
 , protobuf
 , grpc
+, patches ? []
 }:
 
 qtModule {
   pname = "qtgrpc";
   qtInputs = [ qtbase qtdeclarative ];
   buildInputs = [ protobuf grpc ];
+  inherit patches;
 }

